File: functions/services/sns-ranker.js
/**
 * functions/services/sns-ranker.js
 *
 * Twitter 알고리즘 영감의 2단계 랭킹 파이프라인
 * - Light Ranker: flash-lite로 N개 후보 병렬 생성
 * - Heavy Ranker: flash로 최적 후보 선택 (참여도 예측 스코어링)
 *
 * 참고: Twitter/X의 Light Ranker → Heavy Ranker 아키텍처를 SNS 변환에 적용
 *
 * 비용 참고: flash-lite x3 (병렬) + flash x1 (스코어링) = 4 API 호출/플랫폼
 *           기존 flash x2 (순차 retry) 대비 토큰 소비 증가, 품질 상한선 상승
 */

'use strict';

const { callGenerativeModel } = require('./gemini');

// 랭킹 설정 (불변)
const RANKER_CONFIG = Object.freeze({
  candidateCount: 3,              // Light Ranker 후보 수
  lightModel: 'gemini-2.5-flash-lite', // 빠르고 저렴한 모델
  heavyModel: 'gemini-2.5-flash',      // 정밀 평가 모델
  lightTimeoutMs: 20000,          // Light Ranker 개별 타임아웃
  heavyTimeoutMs: 15000,          // Heavy Ranker 타임아웃
  minCandidates: 2,               // 최소 유효 후보 수 (이하면 fallback)
});

/**
 * Promise에 타임아웃을 적용 (timer leak 방지)
 * @param {Promise} promise - 원본 Promise
 * @param {number} ms - 타임아웃 (ms)
 * @param {string} message - 타임아웃 에러 메시지
 * @returns {Promise}
 */
function withTimeout(promise, ms, message) {
  let timer;
  const timeout = new Promise((_, reject) => {
    timer = setTimeout(() => reject(new Error(message)), ms);
  });
  return Promise.race([promise, timeout]).finally(() => clearTimeout(timer));
}

/**
 * Light Ranker: 빠르게 N개 후보를 병렬 생성
 *
 * Twitter의 Light Ranker처럼 빠른 모델로 다수의 후보를 생성하고,
 * 이후 Heavy Ranker에서 정밀 평가하는 구조.
 *
 * @param {string} prompt - SNS 변환 프롬프트
 * @param {number} candidateCount - 생성할 후보 수 (기본 3)
 * @returns {Promise<string[]>} 유효한 후보 텍스트 배열
 */
async function lightRank(prompt, candidateCount = RANKER_CONFIG.candidateCount) {
  const tasks = Array.from({ length: candidateCount }, () =>
    withTimeout(
      callGenerativeModel(prompt, 1, RANKER_CONFIG.lightModel, true, 25000, { temperature: 0.8 }),
      RANKER_CONFIG.lightTimeoutMs,
      'Light Ranker 타임아웃'
    ).catch((err) => {
      console.warn('Light Ranker 후보 생성 실패:', err.message);
      return null;
    })
  );

  const rawCandidates = await Promise.all(tasks);

  const candidates = rawCandidates.filter(
    (c) => c && typeof c === 'string' && c.trim().length > 0
  );

  console.log(`⚡ Light Ranker: ${candidates.length}/${candidateCount}개 후보 생성 완료`);
  return candidates;
}

/**
 * Heavy Ranker용 스코어링 프롬프트 생성
 *
 * Twitter의 Heavy Ranker가 신경망으로 참여도를 예측하듯,
 * LLM이 각 후보의 SNS 성과를 예측 평가한다.
 *
 * 평가 기준 (Twitter 알고리즘 기반):
 * 1. 임팩트 (Hook Quality) - 스크롤을 멈추게 하는 힘
 * 2. 참여 예측 (Engagement Prediction) - 좋아요/RT/댓글 유도력
 * 3. 정보 밀도 (Information Density) - 글자당 전달 정보량
 * 4. 형식 준수 (Format Compliance) - 플랫폼 규격 적합도
 * 5. 원본 충실도 (Source Fidelity) - 원본 메시지 보존도
 *
 * @param {string[]} candidates - 후보 텍스트 배열
 * @param {string} platform - SNS 플랫폼 ('x', 'threads', 'facebook-instagram')
 * @param {string} originalContent - 원본 블로그 원고 (요약본)
 * @param {Object} context - 추가 컨텍스트
 * @param {Object} context.platformConfig - 플랫폼 설정 (글자수 제한 등)
 * @param {Object} context.userInfo - 사용자 정보 (이름, 직책)
 * @returns {string} 스코어링 프롬프트
 */
function buildScoringPrompt(candidates, platform, originalContent, context = {}) {
  const originalSummary = originalContent.length > 500
    ? originalContent.substring(0, 500) + '...'
    : originalContent;

  const candidateBlocks = candidates
    .map((c, i) => `--- 후보 ${i + 1} ---\n${c}\n--- 끝 ---`)
    .join('\n\n');

  const { platformConfig, userInfo } = context;
  const platformName = platformConfig?.name || platform;
  const authorLabel = userInfo ? `${userInfo.name} ${userInfo.position}` : '작성자';

  return `당신은 SNS 콘텐츠 성과 예측 전문가입니다.
아래 원본 블로그 원고를 ${platformName} 플랫폼용으로 변환한 ${candidates.length}개 후보를 평가해주세요.

**작성자:** ${authorLabel}
**원본 원고 (요약):**
${originalSummary}

**후보들:**
${candidateBlocks}

**평가 기준 (가중치 차등, 총 100점):**

1. **임팩트 (Hook Quality)** [25점] - 타임라인에서 스크롤을 멈추게 하는 힘
   - 첫 문장이 관심을 끄는가?
   - 감성적 훅, 질문, 수치, 서사적 대비 등 임팩트 요소가 있는가?
   - 개인 서사나 극적인 숫자 대비가 활용되었는가?

2. **참여 예측 (Engagement Prediction)** [25점] - 좋아요/RT/댓글 유도력
   - 공감을 유도하는 요소가 있는가?
   - CTA(행동 유도)가 자연스러운가?
   - 공유하고 싶은 내용인가?
   - 작성자의 핵심 주제(topic)에 담긴 CTA가 보존되었는가?

3. **정보 밀도 (Information Density)** [20점] - 글자당 전달 정보량
   - 불필요한 수식어 없이 핵심이 전달되는가?
   - 구체적 수치, 고유명사, 사실이 포함되어 있는가?

4. **형식 준수 (Format Compliance)** [15점] - ${platformName} 플랫폼 규격 적합도
   - JSON 형식이 올바른가?
   - 글자수 제한을 준수하는가?${platformConfig?.hashtagLimit ? `\n   - 해시태그 ${platformConfig.hashtagLimit}개 이내인가?` : ''}

5. **원본 충실도 (Source Fidelity)** [15점] - 원본 메시지 보존도
   - 원본의 핵심 메시지가 정확히 전달되는가?
   - 원본에 없는 내용이 추가되지 않았는가?
   - 정치적 입장과 논조가 보존되었는가?

**JSON 출력 형식:**
{
  "rankings": [
    {
      "candidateIndex": 0,
      "scores": {
        "hookQuality": 22,
        "engagementPrediction": 20,
        "informationDensity": 18,
        "formatCompliance": 14,
        "sourceFidelity": 13
      },
      "totalScore": 87,
      "strengths": "서사적 대비 활용, 주제 CTA 보존",
      "weaknesses": "글자수 약간 초과"
    }
  ],
  "bestIndex": 0,
  "reason": "후보 1이 임팩트와 정보 밀도에서 우수"
}`;
}

/**
 * Heavy Ranker: 후보 중 최적 SNS 콘텐츠 선택
 *
 * Twitter의 Heavy Ranker가 참여도를 예측하여 최종 노출 순위를 결정하듯,
 * 이 함수는 LLM으로 각 후보의 SNS 성과를 예측하여 최고 품질 콘텐츠를 선택한다.
 *
 * @param {string[]} candidates - 후보 텍스트 배열
 * @param {string} platform - SNS 플랫폼
 * @param {string} originalContent - 원본 블로그 원고
 * @param {Object} context - platformConfig, userInfo 등 추가 컨텍스트
 * @returns {Promise<{ bestIndex: number, bestCandidate: string, rankings: Array, reason: string }>}
 */
async function heavyRank(candidates, platform, originalContent, context = {}) {
  if (candidates.length === 0) {
    return { bestIndex: -1, bestCandidate: null, rankings: [], reason: '후보 없음' };
  }

  if (candidates.length === 1) {
    return {
      bestIndex: 0,
      bestCandidate: candidates[0],
      rankings: [{ candidateIndex: 0, totalScore: 0 }],
      reason: '단일 후보 (스코어링 스킵)',
    };
  }

  try {
    const scoringPrompt = buildScoringPrompt(candidates, platform, originalContent, context);

    const rawResult = await withTimeout(
      callGenerativeModel(scoringPrompt, 1, RANKER_CONFIG.heavyModel, true, 4096),
      RANKER_CONFIG.heavyTimeoutMs,
      'Heavy Ranker 타임아웃'
    );

    const parsed = parseHeavyRankResult(rawResult, candidates.length);

    console.log(`🏆 Heavy Ranker 결과: 후보 ${parsed.bestIndex + 1} 선택 (${parsed.reason})`);
    if (parsed.rankings.length > 0) {
      parsed.rankings.forEach((r) => {
        console.log(`   후보 ${r.candidateIndex + 1}: ${r.totalScore}점`);
      });
    }

    return {
      ...parsed,
      bestCandidate: candidates[parsed.bestIndex],
    };
  } catch (err) {
    console.warn('🏆 Heavy Ranker 실패, 첫 번째 후보 선택:', err.message);
    return {
      bestIndex: 0,
      bestCandidate: candidates[0],
      rankings: [],
      reason: `Heavy Ranker 실패 (${err.message}), 첫 번째 후보 fallback`,
    };
  }
}

/**
 * Heavy Ranker 결과 파싱
 * @param {string} rawResult - Gemini 응답 텍스트
 * @param {number} candidateCount - 후보 수 (유효성 검증용)
 * @returns {{ bestIndex: number, rankings: Array, reason: string }}
 */
function parseHeavyRankResult(rawResult, candidateCount) {
  try {
    // 균형 잡힌 JSON 추출 (greedy regex 대신 depth 기반 파싱)
    const jsonStr = extractFirstBalancedJson(rawResult);
    if (!jsonStr) {
      throw new Error('JSON 형식 없음');
    }

    const parsed = JSON.parse(jsonStr);
    let bestIndex = parsed.bestIndex;

    // bestIndex 교차 검증: rankings의 최고 점수 후보와 일치하는지 확인
    const rankings = Array.isArray(parsed.rankings) ? parsed.rankings : [];
    if (rankings.length > 0) {
      const highestScored = rankings.reduce((a, b) =>
        (a.totalScore || 0) >= (b.totalScore || 0) ? a : b
      );
      if (typeof highestScored.candidateIndex === 'number' && highestScored.candidateIndex !== bestIndex) {
        console.warn(`bestIndex(${bestIndex})와 최고점수 후보(${highestScored.candidateIndex}) 불일치, 최고점수 기준 보정`);
        bestIndex = highestScored.candidateIndex;
      }
    }

    if (typeof bestIndex !== 'number' || bestIndex < 0 || bestIndex >= candidateCount) {
      throw new Error(`유효하지 않은 bestIndex: ${bestIndex}`);
    }

    return {
      bestIndex,
      rankings,
      reason: parsed.reason || '이유 미제공',
    };
  } catch (err) {
    console.warn('Heavy Ranker 파싱 실패:', err.message);
    return {
      bestIndex: 0,
      rankings: [],
      reason: `파싱 실패 (${err.message}), 첫 번째 후보 fallback`,
    };
  }
}

/**
 * 첫 번째 균형 잡힌 JSON 객체를 추출 (greedy regex 문제 방지)
 * @param {string} text - 원본 텍스트
 * @returns {string|null} JSON 문자열 또는 null
 */
function extractFirstBalancedJson(text) {
  const start = text.indexOf('{');
  if (start === -1) return null;
  let depth = 0;
  let inString = false;
  let escape = false;
  for (let i = start; i < text.length; i++) {
    const ch = text[i];
    if (escape) { escape = false; continue; }
    if (ch === '\\') { escape = true; continue; }
    if (ch === '"') { inString = !inString; continue; }
    if (inString) continue;
    if (ch === '{') depth++;
    else if (ch === '}') depth--;
    if (depth === 0) return text.substring(start, i + 1);
  }
  return null;
}

/**
 * 2단계 랭킹 파이프라인 실행 (메인 함수)
 *
 * Twitter 알고리즘의 Light → Heavy 파이프라인을 SNS 변환에 적용:
 * 1. Light Ranker: flash-lite로 3개 후보 병렬 생성
 * 2. Heavy Ranker: flash로 최고 품질 후보 선택
 *
 * @param {string} prompt - SNS 변환 프롬프트
 * @param {string} platform - SNS 플랫폼
 * @param {string} originalContent - 원본 블로그 원고
 * @param {Object} options - 추가 옵션
 * @param {number} options.candidateCount - 후보 수 (기본 3)
 * @param {Object} options.platformConfig - 플랫폼 설정 (Heavy Ranker 평가에 사용)
 * @param {Object} options.userInfo - 사용자 정보 (Heavy Ranker 평가에 사용)
 * @returns {Promise<{ text: string, ranking: Object }>}
 */
async function rankAndSelect(prompt, platform, originalContent, options = {}) {
  const candidateCount = options.candidateCount || RANKER_CONFIG.candidateCount;

  console.log(`🚀 [SNS Ranker] ${platform} 2단계 랭킹 시작 (후보 ${candidateCount}개)`);
  const startTime = Date.now();

  // Phase 1: Light Rank
  const candidates = await lightRank(prompt, candidateCount);

  if (candidates.length === 0) {
    console.warn('Light Ranker 전체 실패, 기본 모델로 단일 생성 fallback');
    try {
      const fallbackText = await withTimeout(
        callGenerativeModel(prompt, 1, RANKER_CONFIG.heavyModel),
        25000,
        'Fallback 단일 생성 타임아웃 (25초)'
      );
      return {
        text: fallbackText,
        ranking: { bestIndex: 0, rankings: [], reason: 'Light Ranker 전체 실패, 단일 생성 fallback' },
      };
    } catch (err) {
      console.error('Fallback 생성도 실패:', err.message);
      return {
        text: null,
        ranking: { bestIndex: -1, rankings: [], reason: 'Light + Fallback 모두 실패' },
      };
    }
  }

  if (candidates.length < RANKER_CONFIG.minCandidates) {
    console.log(`⚡ 유효 후보 ${candidates.length}개 < ${RANKER_CONFIG.minCandidates}개, Heavy Ranker 스킵`);
    return {
      text: candidates[0],
      ranking: { bestIndex: 0, rankings: [], reason: '후보 부족, Heavy Ranker 스킵' },
    };
  }

  // Phase 2: Heavy Rank (platformConfig, userInfo를 context로 전달)
  const context = {
    platformConfig: options.platformConfig,
    userInfo: options.userInfo,
  };
  const ranking = await heavyRank(candidates, platform, originalContent, context);

  const elapsed = Date.now() - startTime;
  console.log(`🚀 [SNS Ranker] ${platform} 완료: ${elapsed}ms, 후보 ${candidates.length}개 중 #${ranking.bestIndex + 1} 선택`);

  return {
    text: ranking.bestCandidate,
    ranking,
  };
}

module.exports = {
  rankAndSelect,
  lightRank,
  heavyRank,
  buildScoringPrompt,
  withTimeout,
  extractFirstBalancedJson,
  RANKER_CONFIG,
};

File: functions/python/services/sns_ranker.py
"""
SNS 2단계 랭커 (Light -> Heavy).
Node.js `services/sns-ranker.js`의 Python 포팅 버전이다.
"""

from __future__ import annotations

import asyncio
import json
import logging
import time
from typing import Any, Dict, List, Optional

from agents.common.gemini_client import generate_content_async

logger = logging.getLogger(__name__)

RANKER_CONFIG: Dict[str, Any] = {
    "candidateCount": 3,
    "lightModel": "gemini-2.5-flash-lite",
    "heavyModel": "gemini-2.5-flash",
    "lightTimeoutSec": 20,
    "heavyTimeoutSec": 15,
    "minCandidates": 2,
}


async def with_timeout(awaitable, timeout_sec: int | float, message: str):
    try:
        return await asyncio.wait_for(awaitable, timeout=float(timeout_sec))
    except asyncio.TimeoutError as exc:
        raise TimeoutError(message) from exc


def extract_first_balanced_json(text: str) -> Optional[str]:
    if not text:
        return None
    start = text.find("{")
    if start < 0:
        return None

    depth = 0
    in_string = False
    escaped = False

    for idx in range(start, len(text)):
        ch = text[idx]
        if escaped:
            escaped = False
            continue
        if ch == "\\":
            escaped = True
            continue
        if ch == '"':
            in_string = not in_string
            continue
        if in_string:
            continue
        if ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                return text[start : idx + 1]
    return None


async def _generate_light_candidate(prompt: str) -> Optional[str]:
    try:
        text = await with_timeout(
            generate_content_async(
                prompt,
                model_name=RANKER_CONFIG["lightModel"],
                temperature=0.8,
                max_output_tokens=25000,
                response_mime_type="application/json",
                retries=1,
            ),
            RANKER_CONFIG["lightTimeoutSec"],
            "Light Ranker 타임아웃",
        )
        if isinstance(text, str) and text.strip():
            return text
    except Exception as exc:  # pragma: no cover
        logger.warning("Light Ranker 후보 생성 실패: %s", exc)
    return None


async def light_rank(prompt: str, candidate_count: Optional[int] = None) -> List[str]:
    count = candidate_count or RANKER_CONFIG["candidateCount"]
    tasks = [_generate_light_candidate(prompt) for _ in range(count)]
    candidates = await asyncio.gather(*tasks)
    valid = [item for item in candidates if isinstance(item, str) and item.strip()]
    logger.info("Light Ranker: %s/%s개 후보 생성 완료", len(valid), count)
    return valid


def build_scoring_prompt(
    candidates: List[str],
    platform: str,
    original_content: str,
    context: Optional[Dict[str, Any]] = None,
) -> str:
    context = context or {}
    platform_config = context.get("platformConfig") or {}
    user_info = context.get("userInfo") or {}
    source_type = str(context.get("sourceType") or "").strip().lower()
    platform_name = platform_config.get("name", platform)
    author_label = f"{user_info.get('name', '작성자')} {user_info.get('position', '')}".strip()
    source_label = (
        "원본 입장문/페이스북 글"
        if source_type in {"position_statement", "statement", "stance", "facebook_post", "facebook", "fb"}
        else "원본 블로그 원고"
    )

    original_summary = original_content[:500] + "..." if len(original_content) > 500 else original_content
    candidate_blocks = []
    for idx, candidate in enumerate(candidates):
        candidate_blocks.append(f"--- 후보 {idx + 1} ---\n{candidate}\n--- 끝 ---")

    hashtag_rule = ""
    if platform_config.get("hashtagLimit"):
        hashtag_rule = f"\n   - 해시태그 {platform_config.get('hashtagLimit')}개 이내인가?"

    platform_lower = str(platform or "").strip().lower()
    platform_specific_rules = ""
    if platform_lower == "x":
        platform_specific_rules = """
**X 알고리즘 적합도 체크 (필수):**
- 첫 줄 훅이 강하고 스크롤 정지 효과가 있는가?
- 줄바꿈 카드형(2~5줄)으로 모바일 가독성이 확보되는가?
- 블로그 링크가 실제 본문에 포함되는가? (링크 누락 시 감점)
- "자세한 내용은 블로그에서 확인하세요" 같은 저품질 CTA 문구가 없는가? (발견 시 큰 감점)
- 원문의 고유명사/핵심 수치/핵심 주장 중 최소 1개를 보존하는가?
""".strip()
    elif platform_lower == "threads":
        platform_specific_rules = """
**Threads 알고리즘 적합도 체크 (필수):**
- 타래 전체가 훅 -> 맥락 -> 근거 -> 마무리 흐름을 갖는가?
- 각 게시물이 독립적으로도 이해 가능하고, 동시에 이전/다음 게시물과 자연스럽게 이어지는가?
- 게시물 간 문장 반복이 과도하지 않은가? (반복 시 감점)
- 마지막 게시물에 블로그 링크가 포함되는가? (링크 누락 시 감점)
- 과장/선동형 문구 대신 설명형 문체를 유지하는가?
- 원문의 핵심 사실/수치/고유명사를 왜곡 없이 보존하는가?
""".strip()

    return f"""
당신은 SNS 콘텐츠 성과 예측 전문가입니다.
아래 {source_label}를 {platform_name} 플랫폼용으로 변환한 {len(candidates)}개 후보를 평가해주세요.

**작성자:** {author_label}
**원본 원고 (요약):**
{original_summary}

**후보들:**
{chr(10).join(candidate_blocks)}

**평가 기준 (가중치 차등, 총 100점):**

1. **임팩트 (Hook Quality)** [25점]
2. **참여 예측 (Engagement Prediction)** [25점]
3. **정보 밀도 (Information Density)** [20점]
4. **형식 준수 (Format Compliance)** [15점]
   - JSON 형식이 올바른가?
   - 플랫폼 길이 규칙을 준수하는가?{hashtag_rule}
5. **원본 충실도 (Source Fidelity)** [15점]
6. **플랫폼 알고리즘 적합도 (Platform Fit)** [가산/감산]
   - 플랫폼별 요구 문체/구조를 충족하면 가산, 위반 시 감산

{platform_specific_rules}

반드시 JSON으로만 답하세요.
{{
  "rankings": [
    {{
      "candidateIndex": 0,
      "scores": {{
        "hookQuality": 22,
        "engagementPrediction": 20,
        "informationDensity": 18,
        "formatCompliance": 14,
        "sourceFidelity": 13,
        "platformFit": 9
      }},
      "totalScore": 87,
      "strengths": "강점",
      "weaknesses": "약점"
    }}
  ],
  "bestIndex": 0,
  "reason": "선정 이유"
}}
""".strip()


def parse_heavy_rank_result(raw_result: str, candidate_count: int) -> Dict[str, Any]:
    try:
        json_str = extract_first_balanced_json(raw_result)
        if not json_str:
            raise ValueError("JSON 형식 없음")
        parsed = json.loads(json_str)

        rankings = parsed.get("rankings") if isinstance(parsed.get("rankings"), list) else []
        best_index = parsed.get("bestIndex")

        # bestIndex 교차 검증
        if rankings:
            highest = max(rankings, key=lambda item: item.get("totalScore", 0))
            highest_index = highest.get("candidateIndex")
            if isinstance(highest_index, int) and highest_index != best_index:
                logger.warning(
                    "bestIndex(%s)와 최고점수 후보(%s) 불일치, 최고점수 기준으로 보정",
                    best_index,
                    highest_index,
                )
                best_index = highest_index

        if not isinstance(best_index, int) or best_index < 0 or best_index >= candidate_count:
            raise ValueError(f"유효하지 않은 bestIndex: {best_index}")

        return {
            "bestIndex": best_index,
            "rankings": rankings,
            "reason": parsed.get("reason") or "이유 미제공",
        }
    except Exception as exc:
        logger.warning("Heavy Ranker 파싱 실패: %s", exc)
        return {
            "bestIndex": 0,
            "rankings": [],
            "reason": f"파싱 실패 ({exc}), 첫 번째 후보 fallback",
        }


async def heavy_rank(
    candidates: List[str],
    platform: str,
    original_content: str,
    context: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    if not candidates:
        return {"bestIndex": -1, "bestCandidate": None, "rankings": [], "reason": "후보 없음"}
    if len(candidates) == 1:
        return {
            "bestIndex": 0,
            "bestCandidate": candidates[0],
            "rankings": [{"candidateIndex": 0, "totalScore": 0}],
            "reason": "단일 후보 (스코어링 스킵)",
        }

    try:
        scoring_prompt = build_scoring_prompt(candidates, platform, original_content, context=context)
        raw_result = await with_timeout(
            generate_content_async(
                scoring_prompt,
                model_name=RANKER_CONFIG["heavyModel"],
                temperature=0.25,
                max_output_tokens=4096,
                response_mime_type="application/json",
                retries=1,
            ),
            RANKER_CONFIG["heavyTimeoutSec"],
            "Heavy Ranker 타임아웃",
        )
        parsed = parse_heavy_rank_result(raw_result, len(candidates))
        best_index = parsed["bestIndex"]
        return {
            **parsed,
            "bestCandidate": candidates[best_index],
        }
    except Exception as exc:  # pragma: no cover
        logger.warning("Heavy Ranker 실패, 첫 번째 후보 선택: %s", exc)
        return {
            "bestIndex": 0,
            "bestCandidate": candidates[0],
            "rankings": [],
            "reason": f"Heavy Ranker 실패 ({exc}), 첫 번째 후보 fallback",
        }


async def rank_and_select(
    prompt: str,
    platform: str,
    original_content: str,
    options: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    options = options or {}
    candidate_count = options.get("candidateCount") or RANKER_CONFIG["candidateCount"]

    logger.info("[SNS Ranker] %s 2단계 랭킹 시작 (후보 %s개)", platform, candidate_count)
    start_time = time.time()

    candidates = await light_rank(prompt, candidate_count=candidate_count)

    if not candidates:
        logger.warning("Light Ranker 전체 실패, 단일 생성 fallback")
        try:
            fallback_text = await with_timeout(
                generate_content_async(
                    prompt,
                    model_name=RANKER_CONFIG["heavyModel"],
                    temperature=0.25,
                    max_output_tokens=25000,
                    response_mime_type="application/json",
                    retries=1,
                ),
                25,
                "Fallback 단일 생성 타임아웃",
            )
            return {
                "text": fallback_text,
                "ranking": {
                    "bestIndex": 0,
                    "rankings": [],
                    "reason": "Light Ranker 전체 실패, 단일 생성 fallback",
                },
            }
        except Exception as exc:
            logger.error("Fallback 생성도 실패: %s", exc)
            return {
                "text": None,
                "ranking": {
                    "bestIndex": -1,
                    "rankings": [],
                    "reason": "Light + Fallback 모두 실패",
                },
            }

    if len(candidates) < RANKER_CONFIG["minCandidates"]:
        return {
            "text": candidates[0],
            "ranking": {"bestIndex": 0, "rankings": [], "reason": "후보 부족, Heavy Ranker 스킵"},
        }

    ranking = await heavy_rank(
        candidates,
        platform,
        original_content,
        context={
            "platformConfig": options.get("platformConfig"),
            "userInfo": options.get("userInfo"),
        },
    )
    elapsed_ms = int((time.time() - start_time) * 1000)
    logger.info(
        "[SNS Ranker] %s 완료: %sms, 후보 %s개 중 #%s 선택",
        platform,
        elapsed_ms,
        len(candidates),
        ranking.get("bestIndex", -1) + 1,
    )

    return {
        "text": ranking.get("bestCandidate"),
        "ranking": ranking,
    }
